### PR TITLE
[5기] 3주차 Wordle 과제 제출 - 구월

### PR DESCRIPTION
**Pull request by padoling**
### 느낀점
처음 해보는 페어 프로그래밍이었는데, 생각했던 것보다 훨씬 재밌고 유익한 시간이었습니다.
다른 사람이 코드를 짜는 것을 눈 앞에서 지켜보고 제가 코드를 짜는 것도 누군가가 계속 지켜본다는 게 조금 무서웠지만, 막상 해보니 서로의 방식을 따라가보며 많은 것을 배울 수 있어 좋았습니다.
특히 우지님이 TDD나 좋은 객체지향 설계에 대해 많은 의견을 주셔서 더욱 많이 배우고 인사이트를 얻을 수 있었습니다!
기회가 된다면 업무에서도 페어 프로그래밍을 적용해보고 싶습니다.

### 잘한점
- 설계에 긴 시간을 투자하여 후에 큰 변경 없이 코드를 짤 수 있었습니다.
- 테스트 단위별, 기능별로 커밋을 나름 잘 쪼갠 것 같습니다.
- 서로의 생각을 자유롭게 주고받으면서도 무례하지 않게(?) 원활한 커뮤니케이션이 이루어진 것 같습니다.

### 아쉬운점
- 리팩토링을 거치면서 변경된 클래스가 생겼을 때, 테스트 주도 설계의 순서가 약간 꼬이게 된 것 같아 아쉽습니다.
- 제가 속도가 조금 느려서... 더 빨랐다면 리팩토링을 좀 더 해볼 수 있었을텐데 아쉽습니다ㅜㅜ

### 배운점
- TDD에 대해 조금은 알게 되었습니다. 테스트를 위한 객체 설계 방식에 대해서도 배웠습니다.
- 원자적 커밋에 대해 배웠습니다.
- 전체적으로 객체지향적 설계에 대한 지식이 는 것 같습니다.

### CODE REVIEW

File: src/main/java/wordle/domain/Alphabet.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package wordle.domain;
+
+import java.util.Arrays;
+
+public enum Alphabet {
```
Comment (boradol) [inline src/main/java/wordle/domain/Alphabet.java:5]: `Alphabet` enum을 활용 👍

File: src/main/java/wordle/Game.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    public Game(final InputView inputView,
+                final OutputView outputView,
+                final WordListReader wordListReader) {
+        this.inputView = inputView;
+        this.outputView = outputView;
+        this.wordListReader = wordListReader;
+    }
+
+    public void start(final WordSelector wordSelector) {
```
Comment (boradol) [inline src/main/java/wordle/Game.java:25]: 코드가 술술 읽힙니다! 
가독성 정말 좋네요! 👍

File: src/main/java/wordle/Game.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                .boxed()
+                .takeWhile(attempt -> !results.hasAnswer())
+                .forEach(attempt -> {
+                    results.add(examine(wordList, answer));
+                    outputView.showResults(results, attempt, MAX_ATTEMPT);
+                });
+    }
+
+    private Result examine(final WordList wordList, final Answer answer) {
+        final Guess guess = inputWord(wordList);
```
Comment (boradol) [inline src/main/java/wordle/Game.java:44]: `inputWord`메서드에 파라미터에 wordList가 들어가는 것이 조금 부자연스럽게 느껴집니다.

제생각에는 wordList는 GuessWord가 포함되었는지 validation하려고 넘기는 것이라 생각이 드는데요.
뭔가 input이라는 메서드가 다른 행동도 하고 있다 생각이 드는데요.
메서드 명에 대해 어떻게 생각하세용? 🥲

File: src/main/java/wordle/Game.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        return answer.examineResult(guess);
+    }
+
+    private Guess inputWord(final WordList wordList) {
+        try {
+            outputView.insertWord();
+            final Word word = inputView.inputWord();
+            return new Guess(wordList.getWordIfExists(word));
+        } catch (final Exception e) {
```
Comment (boradol) [inline src/main/java/wordle/Game.java:53]: 다시 시도하기 위한 try-catch문이네용!
근데 `Exception`으로 잡게되면, 
다시 시도하고 싶어하는 예외들 말고도 
다른 예외 발생시(IOException 같은...)에도 다시시도가 되지 않을까요?

혹시 다시 시도하고 싶은 경우가 어떤 경우였을까요?🥹

File: README.md [hunk 5]
```diff
@@ -1,89 +1,171 @@
 # 미션 - 워들
-
-## 🔍 진행 방식
-
-- 미션은 **과제 진행 요구 사항**, **기능 요구 사항**, **프로그래밍 요구 사항** 세 가지로 구성되어 있다.
-- 세 개의 요구 사항을 만족하기 위해 노력한다. 특히 기능을 구현하기 전에 기능 목록을 만들고, 기능 단위로 커밋 하는 방식으로 진행한다.
-- **기능 요구 사항에 기재되지 않은 내용은 스스로 판단하여 구현한다.**
+- 참여자 : 우지, 구월
+
+## 시퀀스 다이어그램
```
Comment (boradol) [inline README.md:4]: 시퀀스 다이어그램 포함하느라 고생많으셨어요!
전체적인 맥락이 보여서 정말 좋았어요!!! 👍

File: README.md [hunk 6]
```diff
@@ -1,89 +1,171 @@
...
-    - 맞는 글자는 초록색, 위치가 틀리면 노란색, 없으면 회색
-    - 두 개의 동일한 문자를 입력하고 그중 하나가 회색으로 표시되면 해당 문자 중 하나만 최종 단어에 나타난다.
-- 정답과 답안은 `words.txt`에 존재하는 단어여야 한다.
-- 정답은 매일 바뀌며 ((현재 날짜 - 2021년 6월 19일) % 배열의 크기) 번째의 단어이다.
-
-### 입출력 요구 사항
-
-#### 실행 결과 예시
-
+## 클래스 설계
```
Comment (boradol) [inline README.md:59]: 클래스 설계 부분에서 
View Word Result 이렇게 나눈것을 보고 
크게 이렇게 나눠서 생각한 그림을 볼 수 있었어요.

정말 세세하게 클래스를 잘 나누신것 같아요!
그리고 인터페이스를 적절하게 사용한 부분이 인상적입니다👍

그리고 Word는 어떻게 연관되어있는지 까지 보여주셔서 좋았어요 🥰

File: src/main/java/wordle/domain/Answer.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        return new Result(resultTypes);
+    }
+
+    private ResultType examineResultType(final Guess guess, final int index) {
+        final Alphabet alphabet = guess.find(index);
+        if (alphabet == this.find(index)) {
+            return ResultType.MATCHED;
+        }
+        final long answerCount = countAlphabets(alphabet, size());
+        final long guessCount = guess.countAlphabets(alphabet, Math.addExact(index, EXCLUDE_UNIT));
```
Comment (boradol) [inline src/main/java/wordle/domain/Answer.java:37]: `guessCount`는 두번째 파라미터가 지금 알파벳이 위치하고있는 **(인덱스 + 1)까지의 길이**이고, 
`answerCount`에서는 두번째 파라미터가 어차피 word의 길이 **5**이네요.

제 개인적인 생각으로는 
word에 `countAlphabets(alphabet)` 이런식으로 오버로딩으로 하나 더 만들고 (아님 다른 메서드명으로)
answer 의 `countAlphabets(alphabet, size())`는 두번째 파라미터를 줄일 수 있을 것 같아요.
이 내부에 사용하는 `size()` 함수도 없앨 수 있을것 같아요.

어쨋든 size()함수가 List개수를 또 세고 있는것 같아서용! 
(answer.size()는 여기서만 사용하기도 합니다.)

이건 그냥 제 개인적인 생각입니다.!🤗

File: src/main/java/wordle/domain/Answer.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        if (answerCount >= guessCount) {
```
Comment (boradol) [inline src/main/java/wordle/domain/Answer.java:38]: 저랑 다른 방식의 로직 멋지네요!🤩

해당 알파벳이 정답 단어에서 가지고 있는 총 개수와 
답안이 위치한 알파벳 인덱스까지의 개수로 비교하는 생각 좋네요! 😃

멋집니당!

File: src/main/java/wordle/domain/Results.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        this(new ArrayList<>());
+    }
+
+    public Results(final List<Result> results) {
+        this.results = results;
+    }
+
+    public boolean hasAnswer() {
+        return results.stream()
+                .anyMatch(Result::allMatched);
```
Comment (boradol) [inline src/main/java/wordle/domain/Results.java:21]: 저희 팀도 고민했었던 부분이라 보였어요.🤣

물론 지금 게임 결과에서 그렇게 크리티컬한 부분은 아니지만,
Results의 마지막 Result가 `allMatched`인 것을 판별하는 것이 사실 더 맞지 않을까 생각해봣어요.

혹시 구찌팀도 이 부분에 대해서 고민해보셨었나용? 🤗

File: src/main/java/wordle/domain/ResultType.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package wordle.domain;
+
+public enum ResultType {
+    MATCHED,
+    EXIST,
+    MISMATCHED
```
Comment (boradol) [inline src/main/java/wordle/domain/ResultType.java:6]: 저 저번에 이 이름보고 네이밍이 너무 좋아서 반해버렸습니다.😉

File: src/main/java/wordle/domain/Results.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    public void add(final Result result) {
+        results.add(result);
+    }
+
+    public int size() {
+        return results.size();
+    }
+
+    public List<Result> getResults() {
+        return Collections.unmodifiableList(results);
```
Comment (boradol) [inline src/main/java/wordle/domain/Results.java:33]: `unmodifiableList`👍

여기 뿐만 아니라 코드 전반적으롭 불변성을 지키시려는 노력! 많이 보입니다.
정말 잘 짜여진 코드같습니다.

File: src/main/java/wordle/domain/Word.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    public long countAlphabets(final Alphabet alphabet, final int toIndex) {
+        if (toIndex > alphabets.size()) {
+            throw new IllegalArgumentException("단어의 길이보다 작거나 같은 인덱스만 들어올 수 있습니다");
+        }
+        return alphabets.subList(MINIMUM_INDEX, toIndex)
+                .stream()
+                .filter(it -> it.equals(alphabet))
+                .count();
```
Comment (boradol) [inline src/main/java/wordle/domain/Word.java:47]: 이건 보는 관점에 따라 중요하게 볼 수도 있고 아닐 수도 있지만,
`count()`가 long을 던져주다보니 이것과 관련된 모든 타입이 long이 되더라고요.

고민하고 판단하여 작성하셨을 것 같지만, long까지 할 필요가 있을까라는 생각이 들었어용🤣

File: src/main/java/wordle/domain/WordList.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Objects;
+
+public class WordList {
+    private final List<Word> wordList;
+
+    public WordList(final List<Word> wordList) {
+        this.wordList = wordList;
+    }
+
+    public Word select(final WordSelector wordSelector) {
```
Comment (boradol) [inline src/main/java/wordle/domain/WordList.java:14]: 이부분 인터페이스로 파라미터 넘겨 구현체에 직접 의존하지 않아서 정말 좋았던 부분입니다. 👍

사실 이런 부분이 한두개여야죠!! 👻
진짜 너무 멋진 팀이에요!

File: src/main/java/wordle/view/ConsoleOutputView.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Override
+    public void wrongWord() {
+        System.out.println(WRONG_WORD_MESSAGE);
+    }
+
+    @Override
+    public void showResults(final Results results, final int attempt, final int maxAttempt) {
+        if (results.hasAnswer() || attempt == maxAttempt) {
```
Comment (boradol) [inline src/main/java/wordle/view/ConsoleOutputView.java:35]: 이 부분은 요구사항에 자세하게 나와있는 부분은 아니지만,

`results.hasAnswer()`같은 경우에는 게임의 결과로 치면 성공
`attempt == maxAttempt` 같은 경우에는 게임의 결과의 실패라고 생각이 들어요.
6번 시도를 다 날려버렸으니까요!

그래서 이렇게 수정해보면 어떨까 싶었습니다.
1. `attempt == maxAttempt`를 게임에 실패하는 메서드로 추출하면 어떨까?
2. 아니면, 두 경우(성공, 실패)를 나눠서 다른 출력을 해보면 어떨까?
라는 두가지 방향성을 생각해 보았어요!

하지만, 역시 깔끔하게 잘 구현하신 것 같아요! 💝

File: src/main/java/wordle/view/ResultColor.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
+package wordle.view;
+
+import wordle.domain.ResultType;
+
+import java.util.Arrays;
+
+public enum ResultColor {
```
Comment (boradol) [inline src/main/java/wordle/view/ResultColor.java:7]: 👍
View를 따로 분리해준 부분 너무 인상깊어요!

File: src/test/java/wordle/domain/AlphabetTest.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+
+public class AlphabetTest {
+
+    @Test
+    void 소문자_영단어로_Alphabet을_생성할_수_있다() {
+        assertDoesNotThrow(() -> Alphabet.of('a'));
+    }
+
+    @Test
```
Comment (boradol) [inline src/test/java/wordle/domain/AlphabetTest.java:14]: `@ParameterizedTest` 를 사용한다면,
아래 한줄의 then 절으로 가능합니다. >_< 
아실 것 같지만 활용하면 더 좋을 것 같아 남겨용!

[ParameterizedTest-문서](https://www.baeldung.com/parameterized-tests-junit-5)

File: src/test/java/wordle/domain/AnswerTest.java [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class AnswerTest {
+
+    @Test
+    void 정답은_정답과_동일한_답안이_들어오면_모두_MATCHED인_결과를_반환한다() {
```
Comment (boradol) [inline src/test/java/wordle/domain/AnswerTest.java:12]: 정말 잘 된 로직이라 생각해요!
제가 여러가지 케이스 넣어 검증해봤는데 정말 최고입니다.

그래서 MATCHED말고도 여러가지 테스트 케이스가 있으면,
혹시나 나중에 프로덕션코드가 변경되더라도 무적방어가 가능할 것 같아용!! 🛡️

File: src/test/java/wordle/domain/AnswerTest.java [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+        // when
+        final Result result = answer.examineResult(guess);
+
+        // then
+        assertEquals(result, expectedResult);
+    }
+
+    @Test
+    void 정답은_5글자가_아니면_예외를_발생한다() {
```
Comment (boradol) [inline src/test/java/wordle/domain/AnswerTest.java:27]: Word에서 하는 검증을 Answer, Guess에서도 하고 있는 것 같아요.
근데 실제 validation이 행해지는 곳은 Word라 생각해요!
물론 두번 체크하는 것은 나쁘지 않다고 생각합니다!!

하지만
예를 들어 Answer에서만 잡고싶은 예외보다 Word에서 잡는 예외가 먼저 터질 때,
서로 다른 Exception으로 잡고 있다면, 테스트 케이스가 깨지는 경우를 경험한 적이 생각나더라구용 ㅠㅠㅠㅠ
또는 Word의 Exception을 다른 CustomException으로 바꾸게 되면 다 변경해야할 수도 있습니다.

이런 경우도 있을 것 같아서 코멘트 남기게 되었어요.

하지만 구찌팀에서 잘 판단하여 작성하셨을거라 생각이 들어용!!! >_<

File: src/test/java/wordle/domain/ResultTest.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ResultTest {
+
+    @Test
+    void 모든_ResultType이_Matched라면_매칭_성공_여부를_true로_반환한다() {
+        final Result result = new Result(List.of(ResultType.MATCHED, ResultType.MATCHED));
```
Comment (boradol) [inline src/test/java/wordle/domain/ResultTest.java:14]: Result는 Word의 개수랑 같아야 한다고 생각하는데 어떻게 생각하세요?
저희 팀도 고민했던 부분이라 코멘트 남깁니당!! 🥹

File: src/test/java/wordle/domain/ResultsTest.java [hunk 20]
```diff
@@ -0,0 +1,61 @@
+package wordle.domain;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class ResultsTest {
```
Comment (boradol) [inline src/test/java/wordle/domain/ResultsTest.java:9]: 꼼꼼한 테스트 케이스 멋집니다 👍

File: src/test/java/wordle/domain/WordTest.java [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class WordTest {
+
+    @Test
+    void 게임_단어에_5글자가_아닌_문자열이_들어오면_예외를_발생한다() {
+        // given when then
+        assertThrowsExactly(IllegalArgumentException.class, () -> new Word("abcdef"));
```
Comment (boradol) [inline src/test/java/wordle/domain/WordTest.java:12]: [경계값 분석](https://velog.io/@kimtg1997/Black-Box-%ED%85%8C%EC%8A%A4%ED%8C%85%EC%9D%98-%ED%95%84%EC%9A%94%EC%84%B1%EA%B3%BC-%EB%B0%A9%EB%B2%95#-boundary-value-analysis%EA%B2%BD%EA%B3%84%EA%B0%92-%EB%B6%84%EC%84%9D) 라는 블랙박스 테스트가 있는데요. 

지금 6자에 대한 테스트만 실행하고 있는데요.
5자 경계에 해당하는 주변 값들을 `@ParameterizedTest`와 함께 이용해서 테스트한다면 
더 좋은 테스트가 될 수 있습니다.

링크 걸어두었으니 시간 되시면 함께 봐주세용!!

**Review by boradol — COMMENTED** [review]
안녕하세요. 구월님 >_<
제가 구월님의 리뷰를 맡게 되었네요!

사실 너무나 훌륭한 코드라서 
제가 감히 여기에 리뷰를 남긴다는 것이 잘 완성된 코드에 이상한 리뷰어 빠뜨리기 같은 느낌이었어요
🤣🤣🤣🤣🤣🤣🤣

그래도 저도 또 다른사람의 코드를 보며 공부한다 생각하였고,
저희 팀이 구현하면서 고민했던 부분들도 함께 나눌 수 있겠다 생각하여
열심히 코드를 살펴보았습니다!

일단 우지님이랑 페어 프로그래밍 하실 때 두분다 너무 독보적이셔서
사실 코드 구현부 보고 정말 잘 설계했다 라는 생각이 들었습니다.

좋은점은 일단 구월님께서 워낙 깔끔하게 코드를 구현하시고 클래스 분리를 잘 하시고
객체지향이란 이거다!!! 라는 것을 보여준 코드라 생각해요!
메서드 명이나 딱 직관적이라서 이해하기 쉬운 코드였어요!
그리고 클래스 다이어그램 부분에서도 엄청난 성의가 보여지는 고퀄리티 과제였다고 생각합니다.🎉
구월님 코드 보면서 저의 부족한 코드가 너무 많이 보여서 사실 제가 리뷰를 해도 되나 싶을 정도였습니다.

그리고 제가 코멘트 남긴부분은
1. 의견을 묻고싶거나
2. 제가 궁금해서 의견을 묻고싶은 부분
3. 미비한 실력이지만 얄팍한 지식으로 알고있는 부분에 대해 도움이 될 수 있는 부분
을 추가로 코멘트 남기게 되었어요!

제가 잘 모르고 얘기한 부분이나 
혹시 제 이야기에 의견 공유해주실 부분 있으면 언제든 말씀해주세요!

그리고 구월님의 코드를 리뷰할 수 있어서 영광이었습니다!!!🙇🏻‍♀️
제가 지금 여행중이라서 새벽에 리뷰를 남기느라 조금 정신줄 놓고 코멘트 남긴 부분이 있을수도 있어여ㅠ🫠
그래서 좀 이상한 말 있으면 ㅠㅠ 꼭 말씀해주세요!!

아무튼 페어프로그래밍 하면서 정말 개발 인생으로 재미있고 색다른 경험을 해보게 되엇는데
이렇게 리뷰어까지 해볼 수 있어 감회가 새롭네요!
더구나 제가 굉장히 좋아하는 구월님 것을 리뷰할 수 있어서 좋았습니다.💝

아무튼 긴 글 읽어주셔서 감사합니다.
구월님 마지막으로 3차 과제 하느라 고생 많으셨어요! 우린 이제 짱짱의 길을 걸을것이에요! 🙌

File: README.md (hunk 5 above)
Comment (padoling) [reply README.md:4]: 우지님이 만드시고 제가 살짝 수정만 한 다이어그램입니다!
저도 처음에 보고 화들짝 놀랐답니다 👍

File: src/main/java/wordle/Game.java (hunk 3 above)
Comment (padoling) [reply src/main/java/wordle/Game.java:44]: 확실히 `inputWord` 메소드가 
단어를 입력받는 역할과 입력받은 단어의 validation을 하는 역할 두 가지를 수행하고 있는 것으로 보이네요. 
좋은 지적 감사합니다!
메서드 네이밍을 더 명확하게 하거나, 각 역할별로 메서드를 분리하는 방향으로 고민해서 다시 반영해보겠습니다!

File: src/main/java/wordle/Game.java (hunk 4 above)
Comment (padoling) [reply src/main/java/wordle/Game.java:53]: 처음에 Exception을 명확히 정의하지 않은 상태에서 코드를 짰다가 
Exception을 정의한 후 수정하는 것을 깜빡했네요... 🥲
`Guess`를 생성할 때의 validation을 통과하지 못하거나 `wordList`에 없는 단어가 들어오는 경우에
다시 시도하는 것으로 생각했습니다!
이 부분도 수정하여 반영할게요!

File: src/main/java/wordle/domain/Alphabet.java (hunk 1 above)
Comment (padoling) [reply src/main/java/wordle/domain/Alphabet.java:5]: 우지님의 의견이었습니다 👍👍

File: src/main/java/wordle/domain/Answer.java (hunk 7 above)
Comment (padoling) [reply src/main/java/wordle/domain/Answer.java:37]: 저도 슬쩍(?) 고민했던 부분인데, 역시나 지나치지 않고 짚어주시는 보라돌님👍
오버로딩으로 메서드를 하나 더 만들어보겠습니다!

말씀하신대로 `answer.size()`는 여기서만 사용하기도 하고, 그럼에도 public으로 지정되어 있었네요!
과감히 삭제하는 것도 좋은 방법일 것 같습니다ㅎㅎ

File: src/main/java/wordle/domain/Results.java (hunk 9 above)
Comment (padoling) [reply src/main/java/wordle/domain/Results.java:21]: 다들 고민하는 지점이 비슷한게 신기하네요..!ㅋㅋㅋ

저도 마지막 Result만 비교하는 것이 더 명확하다고 생각했지만 지금 방식도 똑같이 잘 작동하기 때문에 일단 뒀는데요,
마지막만 비교하는 방식으로도 한 번 코드를 수정해봐야겠네요.

보라돌님 팀은 결국 어떻게 구현하셨는지 궁금합니다!

File: src/main/java/wordle/domain/Word.java (hunk 12 above)
Comment (padoling) [reply src/main/java/wordle/domain/Word.java:47]: 맞습니다... 처음에는 `count()`에서 반환된 값을 int로 변환하여 반환하도록 구현했는데,
이후 하게 되는 연산이 long 타입으로도 동일하게 이루어지는 연산이기도 하고
오히려 불필요한 형 변환을 하지 않는 것이 좋겠다...라고 생각했던 것 같아요!

이 문제는 어찌 보면 정답은 없는 것 같아요! 하지만 정말 좋은 지적 감사합니다 🙇‍♀️

File: src/main/java/wordle/view/ConsoleOutputView.java (hunk 14 above)
Comment (padoling) [reply src/main/java/wordle/view/ConsoleOutputView.java:35]: 좋은 의견 감사합니다! 한층 코드의 가독성과 책임 분리를 더 잘 챙길 수 있는 제안인 것 같아요 👍

File: src/test/java/wordle/domain/AnswerTest.java (hunk 18 above)
Comment (padoling) [reply src/test/java/wordle/domain/AnswerTest.java:27]: 말씀해주신대로 Answer와 Guess에서 완전히 동일한 내용을 검증하고 있기 때문에
동일한 검증 내용은 Word에서만 진행하면 될 것 같습니다!

예시로 들어주신 Exception 이야기가 굉장히 와닿네요...! 👍

File: src/test/java/wordle/domain/ResultTest.java (hunk 19 above)
Comment (padoling) [reply src/test/java/wordle/domain/ResultTest.java:14]: Result와 Word의 글자 수가 같아야 한다는 말씀이시죠?!
저희도 고민했었는데, 비즈니스 로직 상 Result는 Word에 의해서만 생성되기 때문에
따로 제약을 두지 않았습니다!

이 테스트케이스에서는 개수에 관계없이 모든 ResultType이 Matched인 경우를 테스트하는 것이라서
내부의 ResultType의 개수를 2개로 해도 괜찮다고 생각했던 것 같습니다!
실제 용례를 생각한다면 5개로 하는 것이 좋을 것 같긴 하네요! 👍

File: src/test/java/wordle/domain/WordTest.java (hunk 21 above)
Comment (padoling) [reply src/test/java/wordle/domain/WordTest.java:12]: 너무 좋은 테스트 방법이네요! 참고하겠습니다. 감사합니다🩷

File: src/main/java/wordle/view/ResultColor.java (hunk 15 above)
Comment (padoling) [reply src/main/java/wordle/view/ResultColor.java:7]: 우지님이 색깔 표시는 view쪽 책임으로 두면 좋을 것 같다고 아이디어를 내주셔서
한때 프론트엔드에 몸 담았던 저도 깊게 공감했답니다 🤗
ResultType과 매핑될 수 있는 Enum으로 표현해봤습니다!
인상깊게 봐주셨다니 너무 감사합니다 👍

File: src/test/java/wordle/domain/AlphabetTest.java (hunk 16 above)
Comment (padoling) [reply src/test/java/wordle/domain/AlphabetTest.java:14]: 사실 테스트 극초보인지라 ParameterizedTest를 한 번도 실제로 써 본적은 없었는데요...!
이 기회에 적극 활용해봐야겠네요! 감사합니다👍

File: src/test/java/wordle/domain/AnswerTest.java (hunk 17 above)
Comment (padoling) [reply src/test/java/wordle/domain/AnswerTest.java:12]: 말씀주신대로 MATCHED 이외에 다양한 테스트케이스를 추가해보겠습니다!
핵심 로직에 대한 테스트인지라 상세한 케이스를 추가할수록 불확실성을 줄일 수 있을 것 같아요 👍
